tmt/utils.py

```diff
@@ -3006,3 +3019,259 @@ def wait(
+        were defined in their respective classes.
+        Yields:
+            key names.
+        """
+
+        for keyname, _ in cls._iter_key_annotations():
+            yield keyname
+
+    def items(self) -> Generator[Tuple[str, Any], None, None]:
+        """
+        Iterate over keys and their values.
+        Keys are yielded in the order: keys declared by parent classes first, then
+        keys declared by the class itself, all following the order in which keys
+        were defined in their respective classes.
+        Yields:
+            pairs of key name and its value.
+        """
+
+        for keyname in self.keys():
+            yield (keyname, getattr(self, keyname))
+
+    # TODO: exists for backward compatibility for the transition period. Once full
+    # type annotations land, there should be no need for extra _keys attribute.
+    @classmethod
+    def _keys(cls) -> List[str]:
+        """ Return a list of names of object's keys. """
+
+        return list(cls.keys())
+
+    def _load_keys(
+            self,
+            key_source: Dict[str, Any],
+            key_source_name: str,
+            logger: Common) -> None:
+        """ Extract values for class-level attributes, and verify they match declared types. """
+
+        LOG_SHIFT, LOG_LEVEL = 2, 4
+
+        debug_intro = functools.partial(logger.debug, shift=LOG_SHIFT - 1, level=LOG_LEVEL)
+        debug = functools.partial(logger.debug, shift=LOG_SHIFT, level=LOG_LEVEL)
+
+        debug_intro('key source')
+        for k, v in key_source.items():
+            debug(f'{k}: {v} ({type(v)})')
+
+        debug('')
+
+        for keyname, keytype in self._iter_key_annotations():
+            key_address = f'{key_source_name}:{keyname}'
+            source_keyname = key_to_option(keyname)
+
+            # Do not indent this particular entry like the rest, so it could serve
+            # as a "header" for a single key processing.
+            debug_intro('key', key_address)
+            debug('field', source_keyname)
+
+            debug('desired type', str(keytype))
+
+            value: Any = None
+
+            debug('dict', self.__dict__)
+
+            if hasattr(self, keyname):
+                # If the key exists as instance's attribute already, it is because it's been
+                # declared with a default value, and the attribute now holds said default value.
+                default_value = getattr(self, keyname)
+
+                # If the default value is a mutable container, we cannot use it directly.
+                # Should we do so, the very same default value would be assigned to multiple
+                # instances/attributes instead of each instance having its own distinct container.
+                if isinstance(default_value, (list, dict)):
+                    debug('detected mutable default')
+                    default_value = copy.copy(default_value)
+
+                debug('default value', str(default_value))
```

Comment:
Tested with:

```yaml
test: ./test.sh
require: tree
```

Which gives:

```
key: /example:require
    field: require
    desired type: typing.List[typing.Union[str, tmt.base.FmfId]]
    dict: {'path': '/example', ...
    detected mutable default
    default value: ['tree']
    default value type: <class 'list'>
    raw value: tree
    raw value type: <class 'str'>
    value: tree
    value type: <class 'str'>
    normalized value: ['tree']
    normalized value type: <class 'list'>
    final value: ['tree']
    final value type: <class 'list'>
    ...
```

Should not the default value for `require` key be a `[]` instead of `['tree']`?

Reply:
Yeah, confirmed :/

```
/foo:
  test: ./test.sh
  require: tree

/bar:
  test: ./test.sh
  require: tree
```

`/foo` is fine, `/bar` is affected and sees `['tree']` :/

Reply:
Fixed.

Reply:
Hm, with the minimal two-line example I still get:

```
    default value: ['tree']
    default value type: <class 'list'>
```

Reply:
Hopefully fixed *now*.

Reply:
Confirmed, thanks!